apiserver: return 4xx for invalid patch

nikhita opened the pull request:
Fixes #54423 

Currently, an invalid patch returns 500. The apiserver should return a 400 (`BadRequest`) or 422 (`Unprocessable Entity`).

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

nikhita committed:
apiserver: return 4xx for invalid patch

Add interpretPatchError to return appropriate http code
(400 or 422) according to the error type.

We add this function in apiserver because we don't want
to mention the http code in apimachinery. The apimachinery
code is also used in kubectl. The client should not return
a server error.

Add a test to validate the http error code and error message.

nikhita commented:
/cc @sttts @caesarxuchao

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -104,6 +104,7 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
 		result, err := patchResource(ctx, updateAdmit, timeout, versionedObj, r, name, patchType, patchJS,
 			scope.Namer, scope.Creater, scope.Defaulter, scope.UnsafeConvertor, scope.Kind, scope.Resource, codec)
 		if err != nil {
+			err = errors.NewBadRequest(err.Error())
```
what if there is a sensible http-compatible error from patchResource? Like NotFound.

nikhita commented:
@sttts Updated. PTAL, thanks.

nikhita commented:
/retest

caesarxuchao commented:
Regarding 400 vs. 422, I found this [description](https://restpatterns.mindtouch.us/HTTP_Status_Codes/422_-_Unprocessable_Entity) for 422: "For example, this error condition (422) may occur if an XML request body contains well-formed (i.e., syntactically correct), but semantically erroneous, XML instructions."

So 400 seems more appropriate for ErrBadJSONDoc, ErrBadPatchFormatForPrimitiveList, ErrBadPatchFormatForRetainKeys, ErrBadPatchFormatForSetElementOrderList, and 422 seems more appropriate for ErrNoListOfLists and ErrPatchContentNotMatchRetainKeys. WDYT?

cc @cheftako

cheftako commented:
@caesarxuchao I just checked 422 in https://tools.ietf.org/html/rfc4918#page-78 and I would concur.

nikhita commented:
Updated.

caesarxuchao commented:
lgtm. @sttts?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -204,7 +204,7 @@ func patchResource(
 					originalPatchMap := make(map[string]interface{})
 					if err := json.Unmarshal(originalPatchBytes, &originalPatchMap); err != nil {
-						return nil, err
+						return nil, mergepatch.ErrBadJSONDoc
```
is there no usable information in the Unmarshl error we are dropping with this?

sttts commented on staging/src/k8s.io/apimachinery/pkg/util/mergepatch/errors.go:
```diff
@@ -17,28 +17,31 @@ limitations under the License.
-	ErrBadPatchFormatForRetainKeys          = errors.New("invalid patch format of retainKeys")
-	ErrBadPatchFormatForSetElementOrderList = errors.New("invalid patch format of setElementOrder list")
-	ErrPatchContentNotMatchRetainKeys       = errors.New("patch content doesn't match retainKeys list")
+	ErrBadJSONDoc                           = apierrors.NewBadRequest("invalid JSON document")
```
While technically this works, I have slight problems to define the http error code of apiserver http handlers in some apimachinery util package deep in the call stack. But the alternative using some more abstract error types here and to map them to http errors in the handlers does not look appealing either.

caesarxuchao replied:
Good point. I think client code (kubectl?) also calls the functions in this package. It will be confusing if server errors are returned by client-side code. cc @mengqiy

Maybe adding a InterpretPatchError function in the handler to map the abstract error to http errors?

sttts replied:
with kubectl in mind that's probably better

sttts commented:
Can we add a unit test in patch_test.go for this?

nikhita commented:
Removed the http error code from apimachinery and added a test.

However, this isn't 100% correct. This will always give a 400. To decide whether we should return a 400 or 422 or 500, we can compare the error to the errors defined in apimachinery and return an error code accordingly. But not sure if this is a good way to go.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -186,7 +186,7 @@ func patchResource(
 				originalJS, patchedJS, err := patchObjectJSON(patchType, codec, currentObject, patchJS, objToUpdate, versionedObj)
 				if err != nil {
-					return nil, err
+					return nil, errors.NewBadRequest(err.Error())
```
`if err, ok := err.(mergepatch.ErrBadJSONDoc); ok { ...`

nikhita commented:
Updated. PTAL, thanks.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -456,3 +456,15 @@ func applyPatchToObject(
+
+// interpretPatchError interprets the error type and returns an error with appropriate HTTP code.
+func interpretPatchError(err error) error {
+	switch err {
```
`switch err.(type) {`, no?

sttts replied:
https://tour.golang.org/methods/16

nikhita replied:
`mergepatch.ErrBadJSONDoc`, etc are all of type `error`. So `err.(type)` will give `error` everytime.

Since `mergepatch.ErrBadJSONDoc` is already of type `error`, we need to check the values (instead of a type switch)

sttts replied:
oh, so these are all constant error objects? Then your code is right. :+1:

nikhita replied:
> oh, so these are all constant error objects? 

Yes, they are constant error objects formatted with the given text. :)

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go:
```diff
@@ -98,6 +98,26 @@ func TestPatchAnonymousField(t *testing.T) {
+
+	actual := &testPatchType{}
+	err := strategicPatchObject(codec, defaulter, original, []byte(patch), actual, &testPatchType{})
+	if err.Error() != expectedError {
```
Could you check if it's http error (like [IsNewBadRequest](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L449))and validate the error code?

nikhita replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go:
```diff
@@ -98,6 +98,29 @@ func TestPatchAnonymousField(t *testing.T) {
+
+	actual := &testPatchType{}
+	err := strategicPatchObject(codec, defaulter, original, []byte(patch), actual, &testPatchType{})
+	if apierrors.IsBadRequest(err) == false {
```
!apierrors.IsBadRequest(err)

sttts commented:
/lgtm
